chore(docs): explain constructs in contrast to modules better (#2715)
hashicorp/terraform-cdk: website/docs/cdktf/concepts/constructs.mdx

# Constructs

## What is a construct?

In CDKTF, a construct is a building block that can represent anything from a single resource to a complex subset of your infrastructure. Each CDKTF program can be thought of as a tree being made up of constructs.

Since constructs can also represent a collection of resource, they are often compared with Terraform modules, but constructs provide a superset of what Terraform modules can do.

## Various use cases of constructs

- Terraform modules are fixed in the sense that they can generate resources based on hcl configuration, but constructs are not constrained to just creating, and can also modify, enrich, and validate resources created outside of the construct.

> - Constructs can can build complex resources that can be used as any other provider supplied resource without using a different syntax like in Terraform.

Author: I think we mean different than your programming language of choice

> - Constructs allow strict type checking and parameter validations for creating resources. This is not possible in Terraform modules.

Author: yes

- Terraform modules are fixed in the sense that they can generate resources based on hcl configuration, but constructs are not constrained to just creating, and can also modify, enrich, and validate resources created outside of the construct.

- **L1** constructs define single resources or very small units of configuration. For example, the code bindings that CDKTF generates for each Terraform provider are considered L1 constructs. Another example would be creating a custom L1 construct that defines the configuration for an Azure virtual machine.
- Constructs can can build complex resources that can be used as any other provider supplied resource without using a different syntax like in Terraform.

- **L2** constructs define resources and include an intent-driven API with additional helper methods, properties, and functionality. For example, you could create a custom L2 construct that contains a method for adding files to an S3 Bucket and a method for granting resource access to a particular user group.
- Constructs allow strict type checking and parameter validations for creating resources. This is not possible in Terraform modules.

> - **L3** constructs define common design patterns and larger pieces of functionality. For example, you could create a custom L3 construct that configures all of the necessary resources to deploy and host a static website frontend.

Author: Ok, I'll add them towards the end then

> - **L3** constructs define common design patterns and larger pieces of functionality. For example, you could create a custom L3 construct that configures all of the necessary resources to deploy and host a static website frontend.

Reviewer: Sounds good! Maybe in a "Terminology" section?

The following example extends the AWS S3 Bucket resource to add required tags. Any consumer of this `TaggedS3Bucket` can be assured that the correct tags are applied to their resource.

> ```ts

Author: Yes, feel free to create these issues 👍

class TaggedS3Bucket extends S3Bucket {
  constructor(scope: Construct, name: string, config: S3BucketConfig) {
    super(scope, name, config);

    this.tagsInput = {
      ...this.tagsInput,
      owner: "my-team",
      purpose: "storage",
    };
  }
}

class MyStack extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);

    new TaggedS3Bucket(this, "my-bucket", {
      bucket: "my-bucket",
    });
  }
}
```

Constructs are not limited to modifying specific resources and can be used to operate on entire stacks. The following example is a construct that can be used to list all S3 buckets in a stack and output them as a list.

```ts
import { Construct } from "constructs";
class OutputAllBucketNames extends Construct {
  constructor(scope: Construct, name: string) {
    super(scope, name);

    const buckets = [];

>     Aspects.of(TerraformStack.of(this)).add({

Author: Yes, sounds good as a follow up, feel free to create the issue 👍

In the following example, we have a contrived `Cluster` construct that allows various webservices to be registered with it. During development, any consumer of the Cluster class will be able to see that any `Cluster` construct accepts `Webservice` constructs to be connected to it.

```ts
import { Construct } from "constructs";

class Cluster extends Construct {
  constructor(scope: Construct, name: string) {
    super(scope, name);
  }

  addService(webservice: Webservice) {
    // e.g. register a service in a k8s cluster
  }
}

class Webservice extends Construct {
  constructor(scope: Construct, name: string) {
    super(scope, name);
  }
}

class MyStack extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);

    const cluster = new Cluster(this, "cluster");
    const webservice = new Webservice(this, "webservice");

    cluster.addService(webservice);
  }
}
```

### You can codify architectural patterns

> If you have complex architectural patterns you can codify them and have an abstraction where you can create a flexible architecture with less work. In this DataPipeline example you can see that each of the steps in the pipeline are bringing their own logic on how to connect to the next step and how to write their output into an S3Bucket. This means as long as a set of resources can fulfil this contract one could build a pipeline using e.g. LambdaFunctions, Ec2 Instances, and CloudwatchEvents if one implemented the `PipelineStep` class for these types.

Reviewer: ```suggestion
If you have complex architectural patterns you can codify them and have an abstraction where you can create a flexible architecture with less work. In this DataPipeline example you can see that each of the steps in the pipeline is bringing their own logic on how to connect to the next step and how to write their output into an S3Bucket. This means as long as a set of resources can fulfill this contract one could build a pipeline using e.g. LambdaFunctions, EC2 Instances, and CloudwatchEvents if one implemented the `PipelineStep` class for these types.
```

```ts
abstract class PipelineStep extends Construct {
  abstract connect(next: PipelineStep);
  abstract finalize(sink: S3Bucket);
}

class DataPipeline extends Construct {
  _nextStep: PipelineStep;

  constructor(scope: Construct, name: string) {
    super(scope, name);
    // this is always the first step
    this._nextStep = new PipelineWebhookStep(this);
  }

  public addStep(step: PipelineStep) {
    this._nextStep.connect(step);
    this.addAccessPermissions(this._nextStep, step);
    this._nextStep = step;
  }

  public finalize() {
    this._nextStep.finalize();
  }
}

class MyStack extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    const pipeline = new DataPipeline(this, "my-pipeline");
    pipeline.addStep(new TransformInputPipelineStep(this));
    pipeline.addStep(new AggregateValuesPipelineStep(this));
    pipeline.addStep(new PipelineStoreResultStep(this));
    pipeline.finalize();
  }
}
```

### You can codify security constraints earlier in the process

> You can improve the feedback cycles about security concerns for developers from tools later in the workflow like code review or CI solutions like sentinel to the point where the cofiguration is authored. This can be done through type hints (disallowing options that are not save on the type system level) or through validation. Validations allow dynamic verifications and are shown to users as hints during `cdktf synth`, if they are on the error level they also fail the synth.

Reviewer: ```suggestion
You can improve the feedback cycles about security concerns for developers from tools later in the workflow like code review or CI solutions like Sentinel to the point where the configuration is authored. This can be done through type hints (disallowing options that are not safe on the type system level) or through validation. Validations allow dynamic verification and are shown to users as hints during `cdktf synth`, if they are on the error level they also fail the synth.
```

  Micro = "t3.micro"
  Small = "t3.small"
  Medium = "t3.medium"
}
interface Ec2InstanceConfig extends GeneratedEc2InstanceConfig {
  instanceSize: AllowedInstanceSizes;
}

class Ec2Instance extends Construct {
  constructor(scope: Construct, name: string, config: Ec2InstanceConfig) {
    super(scope, name, config);
  }
}
```

#### Through Validations

```ts
import { Construct } from "constructs";

class S3Bucket extends Construct {
  constructor(scope: Construct, name: string, config: S3BucketConfig) {
    super(scope, name);

    if (
      this.isOpenToThePublic(config) &&
      !config.tags["reason-why-this-is-public"]
    ) {
      Annotation.of(this).addWarning(
        `This S3Bucket is open to the public and no reason why is stated in the tags, please add a "reason-why-this-is-public" tag`
      );
    }
  }
}
```

### Constructs and Testing

Since Constructs can contain logic and are not just a data structure, they can be tested. This allows you to test the composition of your constructs and the logic they contain. This can be done with snapshot testing or unit testing.

```ts
import { Testing } from "cdktf";
import MyApplicationsAbstraction from "../main";

> import { Ec2 } from "";

Reviewer: ```suggestion
import { Ec2 } from "./custom-ec2-construct";
```

> describe("Construct is of correct compostion and plans successfully", () => {

Reviewer: ```suggestion
describe("Construct is of correct composition and plans successfully", () => {
```

    // make assertions about the composition of constructs
    it("has EC2 of correct size", () => {
        expect(
            Testing.synthScope((scope) => {
                new MyApplicationsAbstraction(scope, "app", {});

>             }).toHaveResourceWithProperties(Ec2, {

Reviewer: the matcher seems to be called on synthScope() and not on expect()?

Author: Oh yeah, I'll change that

                instanceSize: "t3.micro",
            })
        );
    });
  // assert that plan runs sucessfully before a deploy
  it("successfully plans", () => {
    const app = Testing.app();
    const stack = new TerraformStack(app, "test");
    const myAppAbstraction = new MyApplicationsAbstraction(stack, "app", {});

>     expect(Testing.fullSynth(stack).toPlanSuccessfully);

Reviewer: ```suggestion
    expect(Testing.fullSynth(stack)).toPlanSuccessfully();
```

# Constructs

## What is a construct?

In CDKTF, a construct is a building block that can represent anything from a single resource to a complex subset of your infrastructure. Each CDKTF program can be thought of as a tree being made up of constructs.

> Since constructs can also represent a collection of resource, they are often compared with Terraform modules, but constructs provide a superset of what Terraform modules can do.

Reviewer 2: Minor: `resources`

## Various use cases of constructs

- Terraform modules are limited in the sense that they can generate resources based on HCL configuration, but constructs are not constrained to just creating, and can also modify, enrich, and validate resources created outside of the construct.

- Constructs can build complex resources that can be used as any other provider supplied resource without using a different syntax like in Terraform.

- Constructs allow strict type checking and parameter validation for creating resources. This is not possible in Terraform modules.

- Constructs can be used to codify best practices across the organization / codebase.

- Constructs can be tested like any other code within the CDKTF supported languages.

### Constructs can enrich and modify resource configuration to match best practices

> While Terraform modules can be used to generate a part of infrastructure, constructs can be used to enrich and modify existing resource configurations. This enables authors to codify best practices within constructs and users of those constructs can be assured that their infrastructure is compliant.

Reviewer 2: Minor: `best practices within constructs,` (add comma)
